<a href="https://colab.research.google.com/github/nimu77/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [0]:
### YOUR CODE STARTS HERE
import pandas as pd
import numpy as np

In [2]:

column_names = ['party', 'handicapped_infants', 'water_project_cost_sharing', 'adoption_of_budget_resolution', 'physician_ fee_freeze', 'el_salvador_aid', 
                'religious_groups_in_schools', 'anti_satellite_test_ban', 'aid_to_nicaraguan_contras', 'mx_missile', 'immigration', 'synfuels_corporation_cutback', 
                'education_spending', 'superfund_right_to_sue', 'crime', 'duty_free_exports', 'export_administration_act_southafrica']

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=column_names)
print(df.shape)
df.head()

(435, 17)


party  ... export_administration_act_southafrica
0  republican  ...                                     y
1  republican  ...                                     ?
2    democrat  ...                                     n
3    democrat  ...                                     y
4    democrat  ...                                     y

[5 rows x 17 columns]

### Replacing **"y"** with 1, **"n"** with 0, and **"?"** with np.NaN.

In [3]:
df = df.replace({'y': 1, 'n': 0, '?': np.NAN})
df.head()

party  ...  export_administration_act_southafrica
0  republican  ...                                    1.0
1  republican  ...                                    NaN
2    democrat  ...                                    0.0
3    democrat  ...                                    1.0
4    democrat  ...                                    1.0

[5 rows x 17 columns]

In [0]:
# dividing dataframe into two classes
dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

In [5]:
dem.head()

party  ...  export_administration_act_southafrica
2  democrat  ...                                    0.0
3  democrat  ...                                    1.0
4  democrat  ...                                    1.0
5  democrat  ...                                    1.0
6  democrat  ...                                    1.0

[5 rows x 17 columns]

In [0]:
from scipy import stats

### Using hypothesis testing to find an issue that democrats support more than republicans with p < 0.01

1. **null hypothesis**: sample mean of democrats supporting duty free exports == sample mean of republicans supporting duty free exports.
2. **alternative hypothesis**: sample mean of democrats is higher than republicans supporting duty free exports.
3. **confidence level**= 99% (0.1 threshold)

In [10]:
stats.ttest_ind(dem['duty_free_exports'], rep['duty_free_exports'], nan_policy='omit')

Ttest_indResult(statistic=12.853146132542978, pvalue=5.997697174347365e-32)

- From the test, we can assume to reject the null hypothesis and say democrats supporting duty free exports is higher than the percentage of republicans.

In [11]:
dem['duty_free_exports'].mean()

0.6374501992031872

In [12]:
rep['duty_free_exports'].mean()

0.08974358974358974

### Using hypothesis testing, to find an issue that republicans support more than democrats with p < 0.01.

1. **null hypothesis**: sample mean of democrats and republicans supporting el salvador aid is equal.
2. **alternative hypothesis**: sample mean of republicans supporting for elsalvador aid is higher than democrats.
3. **confidence level**: 99% (0.01 threshold)

In [7]:
stats.ttest_ind(dem['el_salvador_aid'], rep['el_salvador_aid'], nan_policy='omit')

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

- From the test, we can say that we can reject the **null hypothesis**  and also say that republicans support more than democrats for the elsalvador aid.

In [8]:
dem['el_salvador_aid'].mean()

0.21568627450980393

In [9]:
rep['el_salvador_aid'].mean()

0.9515151515151515

### Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

1. **null hypothesis**: sample mean of both democrats and republicans supporting **water project cost sharing** is equal.
2. **alternative hypothesis**: sample of both party supporting the agenda is not equal and is different.
3. **confidence level**: 90% (0.1 threshold)

In [17]:
stats.ttest_ind(dem['water_project_cost_sharing'], rep['water_project_cost_sharing'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

- The result say that we **fail to reject the null hypothesis** and say that there is not much difference between democrats and republicans supporting **water project cost sharing**.

In [18]:
dem['water_project_cost_sharing'].mean()

0.502092050209205

In [19]:
rep['water_project_cost_sharing'].mean()

0.5067567567567568

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!

### One sample t-test

In [20]:
df.head()

party  ...  export_administration_act_southafrica
0  republican  ...                                    1.0
1  republican  ...                                    NaN
2    democrat  ...                                    0.0
3    democrat  ...                                    1.0
4    democrat  ...                                    1.0

[5 rows x 17 columns]

In [22]:
dem['religious_groups_in_schools'].value_counts(dropna=False)

0.0    135
1.0    123
NaN      9
Name: religious_groups_in_schools, dtype: int64

In [24]:
dem_supporting_rel = dem[['party', 'religious_groups_in_schools']]
dem_supporting_rel.head()

party  religious_groups_in_schools
2  democrat                          1.0
3  democrat                          1.0
4  democrat                          1.0
5  democrat                          1.0
6  democrat                          1.0

In [27]:
dem_supporting_rel.dropna(inplace=True)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [29]:
dem_supporting_rel.isnull().sum()

party                          0
religious_groups_in_schools    0
dtype: int64

In [31]:
dem_supporting_rel['religious_groups_in_schools'].value_counts(dropna=False)

0.0    135
1.0    123
Name: religious_groups_in_schools, dtype: int64

### One Sample t-test: on religious groups in schools for democrats:
1. **null hypothesis**: 60% supports of democrats supports the religious groups in school.
2. **alternative hypothesis**: Some other than 60% support the religious groups in school.
3. **confidence level**: 95% (0.05 threshold)

$H_0 : \mu = =\bar{x}$ (/mu is .6)

In [32]:
dem_supporting_rel['religious_groups_in_schools'].mean()

0.47674418604651164

In [33]:
dem_supporting_rel['religious_groups_in_schools'].std()

0.5004296407639852

In [36]:
dem_supporting_rel.shape

(258, 2)

### Calculating t-statistics:
- n = 258
- population mean ($\mu$) = 0.6
- sample mean($\bar{x}$) = 0.47674418604651167
- sample standard deviation(s) = 0.5004296407639852

![One Sample T Statistic](https://lh3.googleusercontent.com/proxy/xgoT-32MEk6zEPFQQMTdqK5sy4hBYLL899ChDJMZLGdaQ9R99Zf98XkgTofockfc5ORFOla-JhOmeBnwrUB4IRNOBq4igRRw_HjCq-oJSgWCikYkPZflVEFaEJgLXga0TJ52AYvVMaoiQBNy4bukdFmw)

In [0]:
# calculating it gives us t-statistics of:
# t-test = 3.9561635896
# p-value = 0.00005

#### Checking using scipy

In [34]:
stats.ttest_1samp(dem_supporting_rel['religious_groups_in_schools'], .6)

Ttest_1sampResult(statistic=-3.9561635901847527, pvalue=9.857653395038602e-05)

### Two Sample t-test

In [37]:
adoption_of_budget = df[['party', 'adoption_of_budget_resolution']]
adoption_of_budget.head()

party  adoption_of_budget_resolution
0  republican                            0.0
1  republican                            0.0
2    democrat                            1.0
3    democrat                            1.0
4    democrat                            1.0

In [38]:
adoption_of_budget.isnull().sum()

party                             0
adoption_of_budget_resolution    11
dtype: int64

In [0]:
adoption_of_budget = adoption_of_budget.dropna()

In [40]:
adoption_of_budget.isnull().sum()

party                            0
adoption_of_budget_resolution    0
dtype: int64

In [41]:
adoption_of_budget['adoption_of_budget_resolution'].value_counts()

1.0    253
0.0    171
Name: adoption_of_budget_resolution, dtype: int64

In [0]:
dem_adop_bud = adoption_of_budget[adoption_of_budget['party'] == 'democrat']
rep_adop_bud = adoption_of_budget[adoption_of_budget['party'] == 'republican']

In [46]:
dem_adop_bud.shape

(260, 2)

In [47]:
rep_adop_bud.shape

(164, 2)

In [48]:
# now calculating the sample mean for both parties
dem_adop_bud['adoption_of_budget_resolution'].mean()

0.8884615384615384

In [49]:
rep_adop_bud['adoption_of_budget_resolution'].mean()

0.13414634146341464

In [50]:
# now calculating the sample standard deviation for both parties
dem_adop_bud['adoption_of_budget_resolution'].std()

0.3154048976795528

In [51]:
rep_adop_bud['adoption_of_budget_resolution'].std()

0.34185330768366606

####  Here, 
1. **null hypothesis**: $\mu$democrat == $\mu$republican
2. **alternate hypothesis**: is not equal
3. **confidence level**: 95% (0.05 threshold)

![two sample t-test](https://magoosh.com/statistics/files/2018/04/chapte23.gif)

#### Here, 
- **Sample mean of X1** = 0.89
- **Sample mean of X2** = 0.13
- **S2 for s1**: 0.1024
- **S2  for s2**: 0.1156
- **N1**: 260
- **N2**: 164

- The t- statistics is 23.0303030303
- p-value: 0

### Checking using scipy

In [52]:
stats.ttest_ind(dem_adop_bud['adoption_of_budget_resolution'], rep_adop_bud['adoption_of_budget_resolution'])

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795405602e-77)